docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,158 @@
+  * If the user is not logged into that server, and no username or password is provided, attempt a login using the same behaviour as `spin login`.  If that succeeds, continue to deployment.  If it fails, exit.
+* If the Hippo URL is **not** specified:
+  * If a username or password is specified, error. A URL is required in this case.
+  * If the user is logged in, use the stored URL and token.
+  * If the user is **not** logged in, attempt to log them into a default URL, such as the default address of a local Fermyon platform. If that fails, print usage information and exit.
+
+> QUESTION: If a user is logged into multiple servers, should we:
+> * Default the URL to the _most recent_ login
+> * Provide a way for the user to say which URL is their default
+> * Require the user to provide a URL
+> * Provide an abstraction (e.g. name, config file path) by which the user can switch between servers
+
+> QUESTION: The last section represents a change in behaviour when no flags are provided. Are we happy with this, or should we direct people to `spin login` and not attempt it automatically?
+
+> QUESTION: Is it correct to reject a username/password if the user is logged in to the chosen server, or should we attempt username/password login in this case?
+
+> QUESTION: What happens if the user is logged in to one server and `$HIPPO_URL` is set to another?  What happens if the user is logged in but the `$HIPPO_USERNAME` is set?  (Thinking here of situations where the user is switching between a username/password server and a device-auth one.)  Should we provide a way to "log into" username/password servers, and use that stored state instead of the environment variables?
+
+## Error flows
+
+### Auth failure on deploy
+
+Deployment would fail if the token was incorrect or expired.
+
+> QUESTION: In this case should we simply fail, or should we attempt login?  _Note:_ if we decide to attempt login we _must_ provide a flag to suppress this (for programmatic consumers who have their own login flows).
+
+It would be useful to have an unambiguous way of reporting this for programmatic consumers (e.g. distinguished error code, JSON output option), so they could display login UI or provide a specific prompt.  That may be a bit much to manage in the first drop though.
+
+## `spin login` detailed considerations
+
+`spin login` is a new command and a relatively complex user experience.  This section describes the behaviour in more detail.
+
+### Browser interaction
+
+As part of device auth flow, the user needs to go to the verification URL in their browser and paste in the code.
+
+> QUESTION: Should we attempt to open the browser automatically?
+> * Pro: it's more convenient
+> * Con: it may be unexpected (we have not asked for consent or warned it will happen)
+> * Con: it may work differently across browsers and operating systems, which may require explanation in the documentation
+
+> QUESTION: Should we copy the user code to the clipboard?
```

Comment:
This is a fairly contentious point for me. On one hand, this is incredibly convenient when I want to log in. On the other hand, if I accidentally execute this command, it could wipe my clipboard's content with junk.

Say that I have some important snippet in my clipboard, and I accidentally fat-finger `spin login`. As a result, I've lost the contents of my clipboard.

IMO a `--copy-to-clipboard=false` flag may make sense here.

Reply:
If copy to clipboard is opt _out_, then fat-fingering will still wipe your clipboard.  Only if it's opt _in_ will users know they're about to overwrite their current clip.  And that greatly reduces the convenience because you have to write a whole extra option.

A possible approach would be to have the "please open your browser to..." be interactive, e.g.:

```
Open the Platform device authorization URL in your browser: https://example.com/device-authorization and enter the code: tWOJCFhY
[c] Copy code to clipboard
[b] Open browser
[r] Ready
# user hits "c"
Copied
[b] Open browser
[r] Ready
# user does the dance then hits "r"
Device authorised!
Let the revels commence
```

But that feels like a very heavyweight solution.

On the positive side, we can at least all agree that it's "a contentious issue."

Reply:
`gh` flow for ref:

```
? How would you like to authenticate GitHub CLI? Login with a web browser

! First copy your one-time code: AE4A-1435
Press Enter to open github.com in your browser... 
```

Reply:
I like that flow quite a bit. I'll look and see if we can implement this in time.

Reply:
I'm all in favor of dropping the clipboard copying "feature" altogether.